### ISSUE TITLE
shell_exec(): Unable to execute '"/usr/bin/php5

### ISSUE THREAD

**Issue by danchao** [phpmetrics/PhpMetrics#209]
Hi,

I've installed PhpMetrics v1.9.2 with composer.

When I run phpmetrics against one of my site, it fails.

The command is
phpmetrics --report-html=e-ticket.html e-ticket/

I get this output:
```
root@redmine:/home/danchao/work# phpmetrics --report-html=e-ticket_20160324.html e-ticket/
PHPMetrics by Jean-François Lépine <https://twitter.com/Halleck45>

  297/1030 [========>-------------------]  28%PHP Warning:  shell_exec(): Unable to execute '"/usr/bin/php5" -l 'e-ticket/include/tcpdf/fonts/uni
2cid_aj16.php' 2>&1' in /root/.composer/vendor/phpmetrics/phpmetrics/src/Hal/Component/File/SyntaxChecker.php on line 37
file e-ticket/include/tcpdf/fonts/uni2cid_aj16.php is not valid and has been skipped
  298/1030 [========>-------------------]  28%PHP Warning:  shell_exec(): Unable to execute '"/usr/bin/php5" -l 'e-ticket/include/tcpdf/fonts/cid
0ct.php' 2>&1' in /root/.composer/vendor/phpmetrics/phpmetrics/src/Hal/Component/File/SyntaxChecker.php on line 37
file e-ticket/include/tcpdf/fonts/cid0ct.php is not valid and has been skipped
  299/1030 [========>-------------------]  29%PHP Warning:  shell_exec(): Unable to execute '"/usr/bin/php5" -l 'e-ticket/include/tcpdf/fonts/cid
0cs.php' 2>&1' in /root/.composer/vendor/phpmetrics/phpmetrics/src/Hal/Component/File/SyntaxChecker.php on line 37
```
Then I tried again, it can be reproduced by running against [TCPDF](http://www.tcpdf.org/).
```
root@redmine:/home/danchao/work# phpmetrics --report-cli tcpdf
PHPMetrics by Jean-François Lépine <https://twitter.com/Halleck45>

 160/227 [===================>--------]  70%PHP Warning:  shell_exec(): Unable to execute '"/usr/bin/php5" -l 'tcpdf/fonts/uni2cid_aj16.php' 2>&1
' in /root/.composer/vendor/phpmetrics/phpmetrics/src/Hal/Component/File/SyntaxChecker.php on line 37
file tcpdf/fonts/uni2cid_aj16.php is not valid and has been skipped
 161/227 [===================>--------]  70%PHP Warning:  shell_exec(): Unable to execute '"/usr/bin/php5" -l 'tcpdf/fonts/cid0ct.php' 2>&1' in /
root/.composer/vendor/phpmetrics/phpmetrics/src/Hal/Component/File/SyntaxChecker.php on line 37
file tcpdf/fonts/cid0ct.php is not valid and has been skipped
 162/227 [===================>--------]  71%PHP Warning:  shell_exec(): Unable to execute '"/usr/bin/php5" -l 'tcpdf/fonts/cid0cs.php' 2>&1' in /
root/.composer/vendor/phpmetrics/phpmetrics/src/Hal/Component/File/SyntaxChecker.php on line 37
file tcpdf/fonts/cid0cs.php is not valid and has been skipped
```
Everything after is failed. 
Some other information for reference.
```
root@redmine:/root# phpmetrics --version
PhpMetrics, by Jean-François Lépine (https://twitter.com/Halleck45) version v1.9.2
root@redmine:/root# php -v
PHP 5.6.17-0+deb8u1 (cli) (built: Jan 13 2016 09:10:12) 
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2015, by Zend Technologies
root@redmine:/root# lsb_release  -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.3 (jessie)
Release:        8.3
Codename:       jessie
root@redmine:/root# free
             total       used       free     shared    buffers     cached
Mem:       4045580    2062248    1983332     155476      51020     999500
-/+ buffers/cache:    1011728    3033852
Swap:       262140       2060     260080
root@redmine:/root# df
Filesystem     1K-blocks    Used Available Use% Mounted on
/dev/root       98846220 3311784  90497996   4% /
devtmpfs         2020816       0   2020816   0% /dev
tmpfs            2022788       0   2022788   0% /dev/shm
tmpfs            2022788  157132   1865656   8% /run
tmpfs               5120       0      5120   0% /run/lock
tmpfs            2022788       0   2022788   0% /sys/fs/cgroup
root@redmine:/root# dpkg -l |grep -i php5
ii  php5-cli                          5.6.17+dfsg-0+deb8u1         amd64        command-line interpreter for the php5 scripting language
ii  php5-common                       5.6.17+dfsg-0+deb8u1         amd64        Common files for packages built from the php5 source
ii  php5-json                         1.3.6-1                      amd64        JSON module for php5
ii  php5-readline                     5.6.17+dfsg-0+deb8u1         amd64        Readline module for php5
```
The /etc/php5/cli/php.ini is not modified after installation.

Not really sure if this is a bug or a system configuration/setting issue.

**Comment by Halleck45** [comment]
Hi,

thanks your for your report. Not sure indeed that concerns PhpMetrics: the path of your PHP binary is based on the `PHP_BINARY` variable.

What is the output when your run  

+ `/usr/bin/php5 -l e-ticket/include/tcpdf/fonts/uni2cid_aj16.php` ?
+ `which php` ?

**Comment by danchao** [comment]
```
root@redmine:/home/danchao/work# /usr/bin/php -l e-ticket/include/tcpdf/fonts/uni2cid_aj16.php
No syntax errors detected in include/tcpdf/fonts/uni2cid_aj16.php
root@redmine:/home/danchao/work# which php
/usr/bin/php
root@redmine:/home/danchao/work# /usr/bin/php -v
PHP 5.6.17-0+deb8u1 (cli) (built: Jan 13 2016 09:10:12) 
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2015, by Zend Technologies
```

there is no PHP_BINARY in my environment.

**Comment by Halleck45** [comment]
`PHP_BINARY` is not an environment variable, but a PHP constant.

Can you run the following command:

```
php -r "var_dump(PHP_BINARY);"
```

According to your `which php`, result should be `/usr/bin/php`. But this constant looks to be "/usr/bin/php5"...

**Comment by danchao** [comment]
```
danchao@redmine:~$ ls -lF /usr/bin/php*
lrwxrwxrwx 1 root root      21 Mar 28 14:25 /usr/bin/php -> /etc/alternatives/php*
-rwxr-xr-x 1 root root 9024168 Jan 13 09:27 /usr/bin/php5*
danchao@redmine:~$ ls -lF /etc/alternatives/php
lrwxrwxrwx 1 root root 13 Mar 28 14:25 /etc/alternatives/php -> /usr/bin/php5*
```

**Comment by danchao** [comment]
Hi,

I've done some more testing.
And here's what I found.
The failed report is on a system with 4GB memory.
The whole project I'm trying to analyze can be successfully completed if I run phpMetrics on a system with 16GB memory.
Both system has php memory_limit = -1
So all the "shell_exec(): Unable to execute '"/usr/bin/php5"" message seems due to the phpMetrics process allocate all the memory and there's no memory for shell_exec() to fork a new php5 process.

Maybe there can be a better way to indicate the exhaustion of memory.
Thanks for the help, and let's close this issue.  :)

**Comment by johnerck** [comment]
Yes. All "shell_exec(): Unable to execute" were showing up due to a lack of memory allocation on my server. I switched from a 1GB server to a 4GB server and used ini_set('memory_limit','3000M'); to give my PHP script up to 3GBs of memory (before the memory limit was left at the default which for me was 128M). Fixed this self imposed memory limit issue and now my calls to shell_exec are all working as expected. Thanks for the tip.
